src/groupByKey.ts

```diff
@@ -2,6 +2,18 @@ type GroupsMap<T> = {
   [key: string]: T[];
 };
 
-export function groupByKey(items, key) {
-  // write code here;
+export function groupByKey<R extends object>(
+  items: R[],
+  key: keyof R,
+): GroupsMap<R> {
+  const result: GroupsMap<R> = {};
+
+  const valuesAll = items.map((it) => it[key]);
+  const uniqueValues = [...new Set(valuesAll)];
+
+  uniqueValues.forEach((val) => {
+    result[val] = items.filter((item) => item[key] === val);
+  });
```

Comment:
The current implementation has a performance issue due to the use of `map`, `Set`, and `filter` within a loop, leading to a complexity of O(n^2) for large datasets. Consider refactoring the code to use a single `reduce` or a loop to construct the `GroupsMap` object. This would allow you to group the items in a single iteration over the `items` array, improving the efficiency to O(n).